unixfs/src/file/reader.rs

```diff
@@ -0,0 +1,333 @@
+use crate::pb::{FlatUnixFs, PBLink, RangeLinks, UnixFsType};
+use std::convert::TryFrom;
+use std::fmt;
+use std::ops::Range;
+
+use crate::file::{FileError, FileMetadata, FileReadFailed, UnwrapBorrowedExt};
+
+/// Navigates the UnixFs files, which are either:
+///  - single block files which have everything needed to all of the contents
+///  - multi block files which have trees of trees until Raw leaf blocks
+///
+/// The trees can have different shapes but it doesn't really matter for our depth-first approach.
+/// For seeking, the each sub-tree linking node will have blocksizes for the trees representing
+/// which the original file offsets covered by the tree.
+///
+/// A file doesn't know it's name. It only has a name when part of a directory, and then the name
+/// is on a PbLink::Name. With UnixFs the names are always UTF-8. The root CID is not interesting
+/// either: we just need the root block.
+pub struct FileReader<'a> {
+    offset: u64,
+    end: Ending,
+    had_links: bool,
+    links: Vec<PBLink<'a>>,
+    data: &'a [u8],
+    blocksizes: Vec<u64>,
+    metadata: FileMetadata,
+}
+
+impl AsRef<FileMetadata> for FileReader<'_> {
+    fn as_ref(&self) -> &FileMetadata {
+        &self.metadata
+    }
+}
+
+// TODO: this could be Range ... It just seemed there seems to be "two kinds" of endings but in
+// reality these are closer to two kinds of ranges or spans.
+#[derive(Debug)]
+enum Ending {
+    /// The block represented a subtree without actual content
+    TreeCoverage(u64),
+    /// The block repressented a leaf with actual content
+    Chunk(u64),
+}
+
+impl Ending {
+    /// Checks wheter or not the next range is good to be processed next.
+    fn check_is_suitable_next(&self, offset: u64, next: &Range<u64>) -> Result<(), FileError> {
+        match self {
+            Ending::TreeCoverage(cover_end) if next.start <= offset && &next.end > cover_end => {
```

Comment:
suggestion (this and the next branch): `Ending::TreeCoverage(ref cover_end)`, so you can omit the unusual (in comparisons) `&` before `next.end`

Reply:
Removing `&` from `&next.end` wowuld be a type mismatch: "expected 'u64', found '&u64'".

Reply:
Aah sorry, missed the `ref` hint. I think with the rust2018 or the relevant ergonomics changes `cover_end` is already the older `ref cover_end`, otherwise it would be moved with the match. But it's a Copy type, so with the pre-ergonomics change not having the `ref` would work. While I think the primitives should always be comparable regardless borrowed or owned, I think this is the nicest way around this at the moment. To borrow or to dereference, but I don't think either of those makes any difference.